File: Problem/By section/Dynamic Programing/Problem 583/problem_583.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> dp;
    
    int minDistance(string word1, string word2) {
        int size_1 = word1.size();
        int size_2 = word2.size();
        
        if(size_1 == 0) return size_2;
        if(size_2 == 0) return size_1;
        
        dp = vector<vector<int>>(size_1 + 1, vector<int>(size_2 + 1, 0));
        for(int i = 0; i < size_1; i++) {
           for(int j = 0; j < size_2; j++) {
               if(word1[i] == word2[j]) {
                   dp[i + 1][j + 1] = 1 + dp[i][j];
               }else{
                   dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j]);
               }
           } 
        }
        
        return (size_1 + size_2 - 2 * dp[size_1][size_2]);
        
    }
};

int main(int argc, char** argv) {
    Solution s;
    
    return 0;
}

File: Problem/By section/Dynamic Programing/Problem 583/problem_583.java
class Solution {
    // dp[i][j] = the number of delete operations that need to be done in order to make 
    // word1[0,i] and word2[0,j] equal
    // update rule: same as longest substring
    private int[][] dp;
    
    // Time: O(n * m), Space: O(n * m)
    public int minDistance(String word1, String word2) {
        int m = word1.length();
        int n = word2.length();
        
        if(m == 0 && n == 0) return 0;
        
        dp = new int[m + 1][n + 1];
        
        dp[0][0] = 0;
        for(int i = 1; i<= m; i++){
            dp[i][0] = i;
        }
         for(int i = 1; i <= n; i++){
            dp[0][i] = i;
        }
        
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                if(word1.charAt(i - 1) == word2.charAt(j - 1)){
                    dp[i][j] = dp[i - 1][j - 1];
                }else{
                    dp[i][j] = Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1);
                }

            }
        }
        return dp[m][n];
        
    }
}